### ISSUE TITLE
Support excluding `Configuration`s in `DependencyFilter`

### ISSUE THREAD

**Issue by acisternino** [GradleUp/shadow#159]
Hi,

I am working with the [Apache Spark](http://spark.apache.org/) data processing framework and I need to create a shadow Jar for my multi-module application and its dependencies.

Like regular application servers, Spark requires that deployed application do not contain the same libraries that are already provided by the runtime environment. However, the problem is that Spark itself has a huge list of transitive dependencies (> 300).

When producing the shadow Jar we need to remove all of them but the `exclude` directive is not recursive. OTOH our application is complex enough that explicitly including only the stuff we need is non-optimal as well.

It would be great if this plugin would implement something like along this line:

```
configurations {
    provided
}
dependencies {
    provided    'org.apache.spark:spark-core_2.11:1.5.0'
}
shadowJar {
    dependencies {
        exclude( configurations.provided )      // including all transitive dependencies
    }
}
```

**Comment by johnrengelman** [comment]
You can explicitly configure which configurations are included in the shadow jar by setting the `configurations` property of the task. The default convention is to merge `configurations.runtime` which means it merges anything in runtime and in any configuration that it extends.

In your example above, as long as you don't say something like `configurations.runtime.extendsFrom configurations.provided`, then the `provided` libraries will not be merged into the shadow jar.

So the best practice here is to create a separate configuration for the libraries that you don't want to be included in the scope and then add them to the task that need them (i.e. `compileJava`'s classpath).

Alternatively, you can create a configuration that holds the libraries that you want merged, set `shadowJar.configurations = [configurations.myconfig]` and the configuration to any task classpaths that might need it.

Or a 3rd option is to do above (Create a shadow specific configuration) and then add the dependencies to both it and `compile`/`runtime` and you can abstract that away pretty easily -> https://github.com/gradle/gradle/blob/master/gradle/dependencies.gradle

**Comment by johnrengelman** [comment]
Same enhancement as https://github.com/johnrengelman/shadow/issues/166

**Comment by punya** [comment]
It's difficult to express this cleanly as a union because of how transitive dependencies need to be handled. If I have

```gradle
dependencies {
  compileOnly "foo"
  compile "bar"
}
```

where `"foo"` and `"bar"` both transitively depend on `"baz"`, I'm forced to exclude `"baz"` manually (either within `dependencies` or within `shadowJar`). That's where the ability to exclude the whole configuration would be really useful.

Thanks for considering this request.

**Comment by johnrengelman** [comment]
I'm not following. In your example `"foo"` will not be included because it is `compileOnly` scoped. Why if `"bar"` has a dependency on `"baz"`, why would you not want it included in the shadow jar? if you excluded it, your application would likely break at runtime because those classes wouldn't be found.

**Comment by punya** [comment]
@johnrengelman the idea is that at runtime, something external to the jar (a container or framework) provides `"foo"` and all its dependencies (including `"baz"`). Thus, it's desirable to avoid including a duplicate `"baz"` in the shadow jar.

**Comment by johnrengelman** [comment]
Then I would declare that dependency as `compileOnly` scope and exclude it from the `compile` configuration.

```
configuration.compile.exclude group: 'some-group', name: 'some-artifact'

dependencies {
  compileOnly "some-group:some-artifact:1.2.3"

  compile "bar"
}
```

**Comment by johnrengelman** [comment]
But I do see the value. This support will come in the future.

**Comment by punya** [comment]
The challenge with declaring the dependency as `compileOnly` is that you have to manually, individually exclude the transitive dependencies of `"foo"`. It's certainly doable, but it's tedious and error-prone, especially when said dependencies change as you bump versions.

**Comment by johnrengelman** [comment]
if you only need `"bar"` then you can simply turn off it's transitives:

```
dependencies {
  compile("bar") {
    transitive = false
  }
}
```

But then you just flip the problem because if it has some transitives that you need, then you would need to explicitly declare them.

**Comment by Goooler** [comment]
I don't think a `Configurations` of dependencies should be excluded in `DependencyFilter`. If you really need to exclude a bunch of dependencies, you should try out `transitive`, `shadow`, or `compileOnly`.

https://gradleup.com/shadow/publishing/#shadow-configuration-and-publishing

**Comment by Goooler** [comment]
More minimization details could be controlled by yourselves after #1176 gets resolved.
